sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala

```diff
@@ -640,6 +641,201 @@ class CollationSuite extends DatasourceV2SQLBase with AdaptiveSparkPlanHelper {
+          sql(s"create table $tableNameLowercase(a ${check.dataType}) using parquet")
+          check.rows.map(row =>
+            // scalastyle:off caselocale
+            sql(
+              s"""
+                 |insert into $tableNameLowercase(a)
+                 |values(${UTF8String.fromString(row).toLowerCase})
+                 |""".stripMargin)
+            // scalastyle:on caselocale
+          )
+          checkResults(tableNameLowercase)
+        }
+    )
+  }
+
+  test("Join on arrays of collated strings") {
+    abstract class JoinCheck(
+        val leftRows: Seq[String],
+        val rightRows: Seq[String],
+        val resultRows: Seq[Any])
+      extends ArrayCheck
+
+    case class JoinSimpleCheck(
+        override val leftRows: Seq[String],
+        override val rightRows: Seq[String],
+        override val resultRows: Seq[Seq[String]])
+      extends JoinCheck(leftRows, rightRows, resultRows) with ArrayCheckSimple
+
+    case class JoinNestedCheck(
+        override val leftRows: Seq[String],
+        override val rightRows: Seq[String],
+        override val resultRows: Seq[Seq[Seq[String]]])
+      extends JoinCheck(leftRows, rightRows, resultRows) with ArrayCheckNested
+
+    val tablePrefix = "test_join_arr_collated"
+    val tableLeft = s"${tablePrefix}_left"
+    val tableLeftLowercase = s"${tableLeft}_lowercase"
+    val tableRight = s"${tablePrefix}_right"
+    val tableRightLowercase = s"${tableRight}_lowercase"
+
+    Seq(
+      // simple
+      JoinSimpleCheck(
+        leftRows = Seq("array('aaa')"),
+        rightRows = Seq("array('AAA')"),
+        resultRows = Seq(Seq("aaa"))
+      ),
+      JoinSimpleCheck(
+        leftRows = Seq("array('aaa', 'bbb')"),
+        rightRows = Seq("array('AAA', 'BBB')"),
+        resultRows = Seq(Seq("aaa", "bbb"))
+      ),
+      JoinSimpleCheck(
+        leftRows = Seq("array('aaa')", "array('bbb')"),
+        rightRows = Seq("array('AAA')", "array('BBB')"),
+        resultRows = Seq(Seq("aaa"), Seq("bbb"))
+      ),
+      JoinSimpleCheck(
+        leftRows = Seq("array('aaa')", "array('bbb')"),
+        rightRows = Seq("array('AAAA')", "array('BBBB')"),
+        resultRows = Seq()
+      ),
+      // nested
+      JoinNestedCheck(
+        leftRows = Seq("array(array('aaa'))"),
+        rightRows = Seq("array(array('AAA'))"),
+        resultRows = Seq(Seq(Seq("aaa")))
+      ),
+      JoinNestedCheck(
+        leftRows = Seq("array(array('aaa', 'bbb'))"),
+        rightRows = Seq("array(array('AAA', 'BBB'))"),
+        resultRows = Seq(Seq(Seq("aaa", "bbb")))
+      ),
+      JoinNestedCheck(
+        Seq("array(array('aaa'), array('bbb'))"),
+        Seq("array(array('AAA'), array('BBB'))"),
+        Seq(Seq(Seq("aaa"), Seq("bbb")))
+      ),
+      JoinNestedCheck(
+        Seq("array(array('aaa'), array('bbb'))"),
+        Seq("array(array('AAA'), array('CCC'))"),
+        Seq()
+      )
+    ).map((check: JoinCheck) =>
```

Comment:
Better to use `.foreach` here. `map` is usually used to build new collection which is not need in this case.